src/lib/middleware/pageCacheMiddleware.ts

```diff
@@ -0,0 +1,86 @@
+import {
+  // @ts-ignore
+  PAGE_CACHE_ENABLED,
+  // @ts-ignore
+  PAGE_CACHE_TYPES,
+  // @ts-ignore
+  PAGE_CACHE_NAMESPACE,
+  // @ts-ignore
+  PAGE_CACHE_VERSION,
+  // @ts-ignore
+  PAGE_CACHE_EXPIRY_SECONDS,
+} from "../../config"
+import { Request, Response, NextFunction } from "express"
+const cache = require("lib/cache.coffee")
+const runningTests = Object.keys(
+  require("desktop/components/split_test/running_tests.coffee")
+)
+const cacheablePageTypes: string[] = PAGE_CACHE_TYPES.split("|")
+
+// Middleware will `next` and do nothing if any of the following is true:
+//
+// * page cache feature is disabled.
+// * a user is signed in.
+// * this isnt a supported cacheable path (there is an allow-list set in ENV).
+// * the page content is uncached.
+// * the cache errors.
+export const pageCacheMiddleware = (
+  req: Request,
+  res: Response,
+  next: NextFunction
+) => {
+  if (!PAGE_CACHE_ENABLED) return next()
+
+  // Returns true if the page type corresponding to `url` is configured cacheable.
+  const isCacheablePageType = (url: string) => {
+    const pathWithQueryParams = url.split("/")[1]
+    const pageType = pathWithQueryParams.split("?")[0]
+
+    return cacheablePageTypes.includes(pageType)
+  }
+  if (!isCacheablePageType(req.url)) return next()
+
+  // @ts-ignore
+  const hasUser = !!req.user
+  if (hasUser) return next()
+
+  const handleCacheGet = (err, html) => {
+    if (err) return next()
+
+    if (html) return res.send(html)
+
+    return next()
+  }
+
+  // Generate cache key that includes all currently running AB tests and outcomes.
+  const runningTestsCacheKey = runningTests
+    .map(test => {
+      const outcome = res.locals.sd[test.toUpperCase()]
+      return `${test}:${outcome}`
+    })
+    .join("|")
```

Comment:
* Is `runningTests` a consistently-sorted map? I _think_ yes, since it comes from `Object.keys(...running_tests.coffee)`, but I'm not sure if keys get added to that elsewhere.
* This could get long. Long keys [don't seem to affect performance much](https://stackoverflow.com/questions/6320739/does-name-length-impact-performance-in-redis), but do require more storage. Still this is much smaller than the content being cached, so probably fine.

Reply:
We can add a `.sort` on the result of `Object.keys` to make sure!